lib/graphql/persisted_queries/resolver.rb

```diff
@@ -18,19 +18,26 @@ def message
         end
       end
 
-      def initialize(extensions, store, hash_generator_proc)
+      include GraphQL::Tracing::Traceable
+
+      def initialize(extensions, store, hash_generator_proc, tracers: nil)
         @extensions = extensions
         @store = store
         @hash_generator_proc = hash_generator_proc
+        @tracers = tracers || []
       end
 
       def resolve(query_str)
         return query_str if hash.nil?
 
         if query_str
-          persist_query(query_str)
+          trace("persist_query", hash: hash) do
+            persist_query(query_str)
+          end
         else
-          query_str = @store.fetch_query(hash)
+          trace("fetch_query", hash: hash) do
+            query_str = @store.fetch_query(hash)
```

Comment:
Another thing we could do is pass a `metadata` hash to `trace` (instead of `{hash: hash}`)and then populate that as the code executes.  The tracer would be called with that metadata hash that wouldn't include results until after the tracer called yield:

```ruby
def trace(key, value)
  # value = {hash: "sha256placeholder", result: nil}
  yield
  # value = {hash: "sha256placeholder", result: "cache-hit"}
end
```

It seems weird that it would change, however, but we could make that very clear in the docs.

Reply:
Agreed, that might be helpful in the future, and this is the advanced feature anyway 🙂